dartdoc_test/lib/src/logger.dart

```diff
@@ -12,15 +12,75 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import 'dart:io';
+import 'dart:io' as io;
 
-var verbose = false;
+import 'analyzer.dart';
 
-enum LogLevel { debug, info, warning, error }
+final supportsAnsi = io.stdout.hasTerminal && io.stdout.supportsAnsiEscapes;
 
-void log(String message, [LogLevel level = LogLevel.info]) {
-  if (level == LogLevel.debug && !verbose) {
-    return;
+// ANSI color codes
+final _reset = supportsAnsi ? '\u001b[0m' : '';
+final _red = supportsAnsi ? '\u001b[31m' : '';
+final _green = supportsAnsi ? '\u001b[32m' : '';
+final _yellow = supportsAnsi ? '\u001b[33m' : '';
+final _bold = supportsAnsi ? '\u001b[1m' : '';
+
+enum LogLevel {
+  debug(0),
+  info(1),
+  warning(2),
+  error(3);
+
+  const LogLevel(this.value);
+
+  final int value;
+
+  bool operator >(LogLevel other) => index > other.index;
+}
+
+typedef LogFunction = void Function(String message, LogLevel level);
+
+final class Logger {
+  final LogFunction _log;
+  Logger(this._log);
+
+  void info(String message) => _log(message, LogLevel.info);
+  void debug(String message) => _log(message, LogLevel.debug);
+  void warning(String message) => _log(message, LogLevel.warning);
+  void error(String message) => _log(message, LogLevel.error);
+  void log(String message, LogLevel level) => _log(message, level);
+}
```

Comment:
Just a crazy idea, feel free to ignore, but we could also do:
```suggestion
extension Logger on LogFunction {
  void info(String message) => this(message, LogLevel.info);
  void debug(String message) => this(message, LogLevel.debug);
  void warning(String message) => this(message, LogLevel.warning);
  void error(String message) => this(message, LogLevel.error);
  void log(String message, LogLevel level) => this(message, level);
}
```